### PR TITLE
Changes to build to fix creation of wheels.

### LINKED ISSUE #806: Not possible to install Ray from git
I want to install Ray for Python 3.

```
$ apt-get install -y cmake pkg-config python3-dev build-essential autoconf curl libtool libboost-all-dev unzip
$ pip3 install git+https://github.com/ray-project/ray.git@37282330c0ea687fd1b983176dce85731fcf189d#subdirectory=python
```

But this tries to install it for python2.7, and not python3, failing with:

```
    CMake Error at cmake_modules/FindNumPy.cmake:62 (message):
      NumPy import failure:
    
      Traceback (most recent call last):
    
        File "<string>", line 1, in <module>
    
      ImportError: No module named numpy
```

Because numpy is installed only for Python 3.

### PR DESCRIPTION
**Pull request by robertnishihara**
This addresses #843.

**It should also do the following.**
- It also addresses #813 through a hack. On CentOS, Arrow is installed to `$ARROW_HOME/lib64` instead of `$ARROW_HOME/lib`, so in that case we just copy it to `$ARROW_HOME/lib`. This is not the *right* solution.
- It should fix #806. After this PR, it should be possible to install Ray from github using
    ```
    pip install git+https://github.com/ray-project/ray.git#subdirectory=python
    ```
    in the past, this would fail in certain settings (e.g., the installation would compile against the wrong Python). You can test that it works on this PR with
    ```
    pip install git+https://github.com/robertnishihara/ray.git@9a91b9cf037d1affcd18659af1ad2167c3799226#subdirectory=python
    ```
- It also fixes a **flatbuffer version conflict** in which Arrow was sometimes built using flatbuffers 1.6.0 and Ray pulled in flatbuffers 1.7.1. Now on Linux we just compile flatbuffers ourselves and pass that into the Arrow compilation. Note, we're compiling flatbuffers twice (once for Ray and once for Arrow, and it really only needs to be done once).
- We're also **compiling boost** now on Linux (so that we can compile with `-fPIC`, so that we can link it statically into arrow and numbuf.

**Some changes we should consider making.**
- Compile flatbuffers and boost on Mac also.
- Only pull in flatbuffers once and pass the same one into Ray and Arrow.
- Remove installation of boost from the installation instructions.
- In `setup.py`, move `cython` from `install_requires` to `setup_requires` (and maybe `setuptools_scm` as well, why not `numpy` as well?), see https://github.com/apache/arrow/blob/master/python/setup.py.

### CODE REVIEW

**Comment by robertnishihara** [comment]
This might not completely work yet. I just pip installed one of the wheels I built after this change and saw the following error (upon importing Ray).

```
Python 2.7.12 (default, Nov 19 2016, 06:48:10) 
[GCC 5.4.0 20160609] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import ray
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/ubuntu/.local/lib/python2.7/site-packages/ray/__init__.py", line 13, in <module>
    from ray.worker import (register_class, error_info, init, connect, disconnect,
  File "/home/ubuntu/.local/lib/python2.7/site-packages/ray/worker.py", line 23, in <module>
    import pyarrow.plasma as plasma
  File "/home/ubuntu/.local/lib/python2.7/site-packages/ray/pyarrow_files/pyarrow/__init__.py", line 32, in <module>
    from pyarrow.lib import cpu_count, set_cpu_count
ImportError: libboost_system.so.1.60.0: cannot open shared object file: No such file or directory
```

**Comment by robertnishihara** [comment]
Note some relevant discussion here https://issues.apache.org/jira/browse/ARROW-1368.

**Comment by robertnishihara** [comment]
After compiling boost ourselves with fPIC (on Linux), things are more or less working. However, Ray still does not work on CentOS. In particular, the plasma manager crashes as soon as it starts up (there's some assertion failure somewhere in flatbuffers).

**Comment by robertnishihara** [comment]
The failure I'm currently seeing when starting a plasma manager on CentOS (or when using a wheel built on CentOS) is the following (an assertion failure immediately when the plasma manager starts, somewhere in flatbuffers). 

```
plasma_manager: /ray/ray/python/ray/core/flatbuffers_ep-prefix/src/flatbuffers_ep-install/include/flatbuffers/flatbuffers.h:582: uint8_t* flatbuffers::vector_downward::make_space(size_t): Assertion `cur_ >= buf_' failed.
```

Doing this in gbd and typing `bt` strangely gives 

```
(gdb) bt
No stack.
```

**Comment by robertnishihara** [comment]
It looks like the problem was a flatbuffer version mismatch (I'm still a little confused about which flatbuffers library the plasma manager's plasma store client was using though).

Anyway, both Arrow and Ray supposedly use flatbuffers 1.7.1, but the wheels are being built on a docker image that contains flatbuffers 1.6.0 already https://github.com/apache/arrow/blob/master/python/manylinux1/scripts/build_flatbuffers.sh, so that was the version of flatbuffers being used by Arrow (in `/usr/bin/flatc`).

**Comment by robertnishihara** [comment]
cc @rshin, it looks like the problem was caused at least in part by a flatbuffers version mismatch. Ray was building flatbuffers 1.7.1, but Arrow was finding flatbuffers 1.6.0 which was already in the docker image that I was using.

File: python/setup.py [hunk 1]
```diff
@@ -81,12 +81,15 @@ def has_ext_modules(self):
       # The BinaryDistribution argument triggers build_ext.
       distclass=BinaryDistribution,
       install_requires=["numpy",
+                        "cython",
```
Comment (xhochy) [inline python/setup.py]: If you build your wheel correctly, cython should only be a build/setup_requires but not needed for installation.

Comment (robertnishihara) [reply python/setup.py]: Thanks! Trying to fix this in #878.

**Pull request merged**